<img style="display: block; margin-left: auto; margin-right: auto" src="./ccsf-logo.png" width="250rem;" alt="The CCSF black and white logo">

<div style="text-align: center;">
    <h1>Lab 03: Data Types and Extending Tables</h1>
    <em>View the related <a href="https://ccsf.instructure.com" target="_blank">Canvas</a> Assignment page for additional details.</em>
</div>

So far, you've used Python to manipulate numbers and work with table. You probably need more practice with data types to deepen your understanding of how to work with data in Python.

In this lab, you'll first see how to represent and manipulate another fundamental type of data: text.  A piece of text is called a *string* in Python. You'll also see how to work with *arrays* of data, such as all the numbers between 0 and 100 or all the words in the chapter of a book. Lastly, you'll create tables and practice analyzing them with your knowledge of table operations.

Run the following code cell

In [1]:
import numpy as np
import math
from datascience import *

## Text

Programming doesn't just concern numbers. Text is one of the most common data types used in programs. 

Text is represented by a **string value** in Python. The word "string" is a programming term for a sequence of characters. A string might contain a single character, a word, a sentence, or a whole book.

To distinguish text data from actual code, we demarcate strings by putting quotation marks around them. Single quotes (`'`) and double quotes (`"`) are both valid, but the types of opening and closing quotation marks must match. The contents can be any sequence of characters, including numbers and symbols. 

We've seen strings before in `print` statements.  Below, two different strings are passed as arguments to the `print` function.

In [2]:
print("I <3", 'Data Science')

I <3 Data Science


Just as names can be given to numbers, names can be given to string values.  The names and strings aren't required to be similar in any way. Any name can be assigned to any string.

In [3]:
one = 'two'
plus = '*'
print(one, plus, one)

two * two


### Task 01 📍

Yuri Gagarin was the first person to travel through outer space.  When he emerged from his capsule upon landing on Earth, he [reportedly](https://en.wikiquote.org/wiki/Yuri_Gagarin) had the following conversation with a woman and girl who saw the landing:

    The woman asked: "Can it be that you have come from outer space?"
    Gagarin replied: "As a matter of fact, I have!"

The cell below contains unfinished code.  Fill in the `...`s so that it prints out this conversation *exactly* as it appears above.


In [4]:
woman_asking = 'The woman asked:' # SOLUTION
woman_quote = '"Can it be that you have come from outer space?"'
gagarin_reply = 'Gagarin replied:'
gagarin_quote = '"As a matter of fact, I have!"' # SOLUTION

print(woman_asking, woman_quote)
print(gagarin_reply, gagarin_quote)

The woman asked: "Can it be that you have come from outer space?"
Gagarin replied: "As a matter of fact, I have!"


In [5]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct string."
failure_message: "❌ It doesn't seem like you have the correct string. Notice that print() adds a space between the strings you provide as input."
""" # END TEST CONFIG

woman_asking == 'The woman asked:'

True

In [6]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct string."
failure_message: "❌ You'll need to use two different types of quotes, so you can use quotes within a string."
""" # END TEST CONFIG

gagarin_quote == '"As a matter of fact, I have!"'

True

### String Methods

Strings can be transformed using **methods**. Recall that methods and functions are not technically the same thing, but we'll be using them interchangeably for the purposes of this course.

Here's a sketch of how to call methods on a string:

    <expression that evaluates to a string>.<method name>(<argument>, <argument>, ...)
    
One example of a string method is `replace`, which replaces all instances of some part of the original string (or a *substring*) with a new string. 

    <original string>.replace(<old substring>, <new substring>)
    
`replace` returns (evaluates to) a new string, leaving the original string unchanged.
    
Try to predict the output of this example, then run the cell!

In [7]:
# Replace one letter
hello = 'Hello'
print(hello.replace('o', 'a'), hello)

Hella Hello


You can call functions on the results of other functions.  For example, `max(abs(-5), abs(3))` evaluates to 5.  Similarly, you can call methods on the results of other method or function calls.

You may have already noticed one difference between functions and methods - a function like `max` does not require a `.` before it's called, but a string method like `replace` does.

In [8]:
# Calling replace on the output of another call to replace
'train'.replace('t', 'ing').replace('in', 'de')

'degrade'

Here's a picture of how Python evaluates a "chained" method call like that:

<img src="chaining_method_calls.png"/>

#### Task 02 📍

Use `replace` to transform the string `'hitchhiker'` into `'matchmaker'`. Assign your result to `new_word`.


In [9]:
new_word = 'hitchhiker'.replace('hi', 'ma') # SOLUTION
new_word

'matchmaker'

In [10]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct string."
failure_message: "❌ Start with 'hitchhiker' and use .replace() after that."
""" # END TEST CONFIG

new_word == 'matchmaker'

True

There are many more string methods in Python, but most programmers don't memorize their names or how to use them.  In the "real world," people usually just search the internet for documentation and examples. A complete [list of string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) appears in the Python language documentation. [Stack Overflow](http://stackoverflow.com) has a huge database of answered questions that often demonstrate how to use these methods to achieve various ends.

### Converting To and From Strings

Strings and numbers are different *types* of values, even when a string contains the digits of a number. For example, evaluating the following cell causes an error because an integer cannot be added to a string.

<img src="./int_add_str_error.png">

However, there are built-in functions to convert numbers to strings and strings to numbers. Some of these built-in functions have restrictions on the type of argument they take:

|Function |Description|
|-|-|
|`int`|Converts a string of digits or a float to an integer ("int") value|
|`float`|Converts a string of digits (perhaps with a decimal point) or an int to a decimal ("float") value|
|`str`|Converts any value to a string|

Try to predict what data type and value `example` evaluates to, then run the cell.

In [11]:
example = 8 + int("10") + float("8")

print(example)
print("This example returned a " + str(type(example)) + "!")

26.0
This example returned a <class 'float'>!


Suppose you're writing a program that looks for dates in a text, and you want your program to find the amount of time that elapsed between two years it has identified.  It doesn't make sense to subtract two texts, but you can first convert the text containing the years into numbers.

#### Task 03 📍

Finish the code below to compute the number of years that elapsed between `one_year` and `another_year`.  Don't just write the numbers `1618` and `1648` (or `30`); use a conversion function to turn the given text data into numbers.


In [12]:
# Some text data:
one_year = "1618"
another_year = "1648"

# Complete the next line.  Note that we can't just write:
#   another_year - one_year
# If you don't see why, try seeing what happens when you
# write that here.
difference = int(another_year) - int(one_year) # SOLUTION
difference

30

In [13]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct value."
failure_message: "❌ You'll need to convert the strings to numbers. Try using the int() function to convert the strings to integers, so you can subtract the values."
""" # END TEST CONFIG

abs(difference) == 30 or round(abs(difference), 1) == 30.0

True

### Passing strings to functions

String values, like numbers, can be arguments to functions and can be returned by functions. 

The function `len` (derived from the word "length") takes a single string as its argument and returns the number of characters (including spaces) in the string.

Note that it doesn't count *words*. `len("one small step for man")` evaluates to 22, not 5.

#### Task 04 📍

Use `len` to find the number of characters in the long string in the next cell.  Characters include things like spaces and punctuation. Assign `sentence_length` to that number.

(The string is the first sentence of the English translation of the French [Declaration of the Rights of Man](http://avalon.law.yale.edu/18th_century/rightsof.asp).)  


In [14]:
a_very_long_sentence = "The representatives of the French people, organized as a National Assembly, believing that the ignorance, neglect, or contempt of the rights of man are the sole cause of public calamities and of the corruption of governments, have determined to set forth in a solemn declaration the natural, unalienable, and sacred rights of man, in order that this declaration, being constantly before all the members of the Social body, shall remind them continually of their rights and duties; in order that the acts of the legislative power, as well as those of the executive power, may be compared at any moment with the objects and purposes of all political institutions and may thus be more respected, and, lastly, in order that the grievances of the citizens, based hereafter upon simple and incontestable principles, shall tend to the maintenance of the constitution and redound to the happiness of all."
sentence_length = len(a_very_long_sentence) # SOLUTION
sentence_length

896

In [15]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ That seems correct!"
failure_message: "❌ Try using len() with the string as the argument."
""" # END TEST CONFIG

sentence_length == 896

True

## Arrays

Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That's if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that contains the result of multiplying each number in `billions_of_numbers` by .18.  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**. 

### Making arrays

First, let's learn how to manually input values into an array. This typically isn't how programs work. Normally, we create arrays by loading them from an external source, like a data file.

To create an array by hand, call the function `make_array`.  Each argument you pass to `make_array` will be in the array it returns.  Run this cell to see an example:

In [16]:
make_array(0.125, 4.75, -1.3)

array([ 0.125,  4.75 , -1.3  ])

Each value in an array (in the above case, the numbers 0.125, 4.75, and -1.3) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them to names or use them as arguments to functions. For example, `len(<some_array>)` returns the number of elements in `some_array`.

#### Task 05 📍

Make an array containing the numbers 0, 1, -1, $\pi$, and $e$, in that order.  Name it `interesting_numbers`.  

*Hint:* How did you get the values $\pi$ and $e$ in lab 2?  You can refer to them in exactly the same way here.


In [17]:
interesting_numbers = make_array(0, 1, -1, math.pi, math.e) # SOLUTION
interesting_numbers

array([ 0.        ,  1.        , -1.        ,  3.14159265,  2.71828183])

In [18]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you made an array."
failure_message: "❌ You need to make an array. Try using the function make_array."
""" # END TEST CONFIG

type(interesting_numbers) == np.ndarray

True

In [19]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ You have the correct number of values."
failure_message: "❌ It seems like you missed some of the values, you should have 5."
""" # END TEST CONFIG

len(interesting_numbers) == 5

True

In [20]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct values."
failure_message: "❌ You don't seem to have the correct values. Don't forgot that pi and e are part of the math module."
""" # END TEST CONFIG

all(interesting_numbers == np.array([0, 1, -1, math.pi, math.e]))

True

#### Task 06 📍

Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you evaluate `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's just NumPy's extremely cryptic way of saying that the data types in the array are strings.


In [21]:
hello_world_components = make_array("Hello", ",", " ", "world", "!") # SOLUTION
hello_world_components

array(['Hello', ',', ' ', 'world', '!'],
      dtype='<U5')

In [22]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you made an array."
failure_message: "❌ You need to make an array. Try using the function make_array."
""" # END TEST CONFIG

type(hello_world_components) == np.ndarray

True

In [23]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ You have the correct number of values."
failure_message: "❌ It seems like you missed some of the values, you should have 5."
""" # END TEST CONFIG

len(hello_world_components) == 5

True

In [24]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct values."
failure_message: "❌ You don't seem to have the correct values."
""" # END TEST CONFIG

all(hello_world_components == np.array(["Hello", ",", " ", "world", "!"]))

True

####  `np.arange`

Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie"). The package is called `numpy`, but it's standard to rename it `np` for brevity.  You can do that with:

    import numpy as np

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  The line of code `np.arange(start, stop, step)` evaluates to an array with all the numbers starting at `start` and counting up by `step`, stopping **before** `stop` is reached.

Run the following cells to see some examples!

In [25]:
# This array starts at 1 and counts up by 2
# and then stops before 6
np.arange(1, 6, 2)

array([1, 3, 5])

In [26]:
# This array doesn't contain 9
# because np.arange stops *before* the stop value is reached
np.arange(4, 9, 1)

array([4, 5, 6, 7, 8])

#### Task 07 📍

Import `numpy` as `np` and then use `np.arange` to create an array with the multiples of 99 from 0 up to (**and including**) 9999.  (So its elements are 0, 99, 198, 297, etc.)


In [27]:
import numpy as np # SOLUTION
multiples_of_99 = np.arange(0, 9999 + 99, 99)
multiples_of_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

In [28]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you made an array."
failure_message: "❌ You need to make an array. Try using the function make_array."
""" # END TEST CONFIG

type(multiples_of_99) == np.ndarray

True

In [29]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ You have the correct number of values."
failure_message: "❌ It seems like you missed some of the values, you should have 5."
""" # END TEST CONFIG

len(multiples_of_99) == 102

True

In [30]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct values."
failure_message: "❌ You don't seem to have the correct values. Don't forgot that pi and e are part of the math module."
""" # END TEST CONFIG

all(multiples_of_99 == np.arange(0, 9999+99, 99))

True

### Indexing

Let's work with a more interesting dataset.  The next cell creates an array called `population_amounts` that includes estimated world populations in every year from **1950** to roughly the present.  (The estimates come from the US Census Bureau website.)

Rather than type in the data manually, we've loaded them from a file on your computer called `world_population.csv`.  You'll learn how to do that later in this lab!

In [31]:
population_amounts = Table.read_table("world_population.csv").column("Population")
population_amounts

array([2557628654, 2594939877, 2636772306, 2682053389, 2730228104,
       2782098943, 2835299673, 2891349717, 2948137248, 3000716593,
       3043001508, 3083966929, 3140093217, 3209827882, 3281201306,
       3350425793, 3420677923, 3490333715, 3562313822, 3637159050,
       3712697742, 3790326948, 3866568653, 3942096442, 4016608813,
       4089083233, 4160185010, 4232084578, 4304105753, 4379013942,
       4451362735, 4534410125, 4614566561, 4695736743, 4774569391,
       4856462699, 4940571232, 5027200492, 5114557167, 5201440110,
       5288955934, 5371585922, 5456136278, 5538268316, 5618682132,
       5699202985, 5779440593, 5857972543, 5935213248, 6012074922,
       6088571383, 6165219247, 6242016348, 6318590956, 6395699509,
       6473044732, 6551263534, 6629913759, 6709049780, 6788214394,
       6866332358, 6944055583, 7022349283, 7101027895, 7178722893,
       7256490011])

Here's how we get the first element of `population_amounts`, which is the world population in the first year in the dataset, 1950.

In [32]:
population_amounts.item(0)

2557628654

The value of that expression is the number 2557628654 (around 2.5 billion), because that's the first thing in the array `population_amounts`.

Notice that we wrote `.item(0)`, not `.item(1)`, to get the first element.  This is a weird convention in computer science.  0 is called the *index* of the first item.  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

Here are some more examples.  In the examples, we've given names to the things we get out of `population_amounts`.  Read and run each cell.

In [33]:
# The 13th element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population_amounts.item(12)
population_1962

3140093217

In [34]:
# The 66th element is the population in 2015.
population_2015 = population_amounts.item(65)
population_2015

7256490011

You will see an `IndexError` if you try to use an index value that is outside the range of possible values for the array:

<img src="./index_error.png">

Since `make_array` returns an array, we can call `.item(3)` on its output to get its 4th element, just like we "chained" together calls to the method `replace` earlier.

In [35]:
make_array(-1, -3, 4, -2).item(3)

-2

#### Task 08 📍

Set `population_1973` to the world population in 1973, by getting the appropriate element from `population_amounts` using `item`.


In [36]:
population_1973 = population_amounts.item(23) # SOLUTION
population_1973

3942096442

In [37]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ You have the correct value."
failure_message: "❌ It doesn't seem that you have the correct value. Try using .item() after the array name using the correct index. You might be off by 1 on the index."
""" # END TEST CONFIG

population_1973 == 3942096442

True

Arrays are primarily useful for doing the same operation many times, so we don't often have to use `.item` and work with single elements.

### Logarithms

Here is one simple question we might ask about world population:

> How big was the population in *orders of magnitude* in each year?

Orders of magnitude quantify how big a number is by representing it as the power of another number (for example, representing 104 as $10^{2.017033}$). One way to do this is by using the logarithm function. The logarithm (base 10) of a number increases by 1 every time we multiply the number by 10. It's like a measure of how many decimal digits the number has, or how big it is in orders of magnitude.

We could try to answer our question like this, using the `log10` function from the `math` module and the `item` method you just saw:

In [38]:
population_1950_magnitude = math.log10(population_amounts.item(0))
population_1951_magnitude = math.log10(population_amounts.item(1))
population_1952_magnitude = math.log10(population_amounts.item(2))
population_1953_magnitude = math.log10(population_amounts.item(3))
# and so on

But this is tedious and doesn't really take advantage of the fact that we are using a computer.

Instead, NumPy provides its own version of `log10` that takes the logarithm of each element of an array.  It takes a single array of numbers as its argument.  It returns an array of the same length, where the first element of the result is the logarithm of the first element of the argument, and so on.

#### Task 09 📍

Use `np.log10` to compute the logarithms of the world population in every year.  Give the result (an array of 66 numbers) the name `population_magnitudes`.  Your code should be very short.


In [39]:
population_magnitudes = np.log10(population_amounts) # SOLUTION
population_magnitudes

array([ 9.40783749,  9.4141273 ,  9.42107263,  9.42846742,  9.43619893,
        9.44437257,  9.45259897,  9.46110062,  9.4695477 ,  9.47722498,
        9.48330217,  9.48910971,  9.49694254,  9.50648175,  9.51603288,
        9.5251    ,  9.53411218,  9.54286695,  9.55173218,  9.56076229,
        9.56968959,  9.57867667,  9.58732573,  9.59572724,  9.60385954,
        9.61162595,  9.61911264,  9.62655434,  9.63388293,  9.64137633,
        9.64849299,  9.6565208 ,  9.66413091,  9.67170374,  9.67893421,
        9.68632006,  9.69377717,  9.70132621,  9.70880804,  9.7161236 ,
        9.72336995,  9.73010253,  9.73688521,  9.74337399,  9.74963446,
        9.75581413,  9.7618858 ,  9.76774733,  9.77343633,  9.77902438,
        9.7845154 ,  9.78994853,  9.7953249 ,  9.80062024,  9.80588805,
        9.81110861,  9.81632507,  9.82150788,  9.82666101,  9.83175555,
        9.83672482,  9.84161319,  9.84648243,  9.85132122,  9.85604719,
        9.8607266 ])

In [40]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you made an array."
failure_message: "❌ You need to make an array. Try using the function make_array."
""" # END TEST CONFIG

type(population_magnitudes) == np.ndarray

True

In [41]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct values."
failure_message: "❌ You don't seem to have the correct values. Make sure you are using np.log10"
""" # END TEST CONFIG

sum(abs(population_magnitudes - np.log10(population_amounts))) < 1e-6

True

What you just did is called *elementwise* application of `np.log10`, since `np.log10` operates separately on each element of the array that it's called on. Here's a picture of what's going on:

<img src="array_logarithm.jpg">


The textbook's [section](https://www.inferentialthinking.com/chapters/05/1/Arrays)  on arrays has a useful list of NumPy functions that are designed to work elementwise, like `np.log10`.

### Arithmetic

Arithmetic also works elementwise on arrays, meaning that if you perform an arithmetic operation (like subtraction, division, etc) on an array, Python will do the operation to every element of the array individually and return an array of all of the results. For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [42]:
population_in_billions = population_amounts / 1000000000
population_in_billions

array([ 2.55762865,  2.59493988,  2.63677231,  2.68205339,  2.7302281 ,
        2.78209894,  2.83529967,  2.89134972,  2.94813725,  3.00071659,
        3.04300151,  3.08396693,  3.14009322,  3.20982788,  3.28120131,
        3.35042579,  3.42067792,  3.49033371,  3.56231382,  3.63715905,
        3.71269774,  3.79032695,  3.86656865,  3.94209644,  4.01660881,
        4.08908323,  4.16018501,  4.23208458,  4.30410575,  4.37901394,
        4.45136274,  4.53441012,  4.61456656,  4.69573674,  4.77456939,
        4.8564627 ,  4.94057123,  5.02720049,  5.11455717,  5.20144011,
        5.28895593,  5.37158592,  5.45613628,  5.53826832,  5.61868213,
        5.69920299,  5.77944059,  5.85797254,  5.93521325,  6.01207492,
        6.08857138,  6.16521925,  6.24201635,  6.31859096,  6.39569951,
        6.47304473,  6.55126353,  6.62991376,  6.70904978,  6.78821439,
        6.86633236,  6.94405558,  7.02234928,  7.10102789,  7.17872289,
        7.25649001])

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`). For example, you can calculate a tip on several restaurant bills at once (in this case just 3):

In [43]:
restaurant_bills = make_array(20.12, 39.90, 31.01)
print("Restaurant bills:\t", restaurant_bills)

# Array multiplication
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [ 20.12  39.9   31.01]
Tips:			 [ 4.024  7.98   6.202]


<img src="array_multiplication.jpg">

#### Task 10 📍

Suppose the total charge at a restaurant is the original bill plus the tip. If the tip is 20%, that means we can multiply the original bill by 1.2 to get the total charge.  Compute the total charge for each bill in `restaurant_bills`, and assign the resulting array to `total_charges`.


In [44]:
total_charges = 1.2 * restaurant_bills # SOLUTION
total_charges

array([ 24.144,  47.88 ,  37.212])

In [45]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you made an array."
failure_message: "❌ You need to make an array. Try using the function make_array."
""" # END TEST CONFIG

type(total_charges) == np.ndarray

True

In [46]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct values."
failure_message: "❌ You don't seem to have the correct values. Try multiplying the array by 1.2"
""" # END TEST CONFIG

total_charges.item(0) / 1.2 == restaurant_bills.item(0) \
and total_charges.item(2) / 1.2 != restaurant_bills.item(2)

True

#### Task 11 📍

The array `more_restaurant_bills` contains 100,000 bills!  Compute the total charge for each one.  How is your code different?


In [47]:
more_restaurant_bills = Table.read_table("more_restaurant_bills.csv").column("Bill")
more_total_charges = 1.2 * more_restaurant_bills # SOLUTION
more_total_charges

array([ 20.244,  20.892,  12.216, ...,  19.308,  18.336,  35.664])

In [48]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you made an array."
failure_message: "❌ You need to make an array. Try using the function make_array."
""" # END TEST CONFIG

type(more_total_charges) == np.ndarray

True

In [49]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct values."
failure_message: "❌ You don't seem to have the correct values. Try multiplying the array by 1.2"
""" # END TEST CONFIG

sum(abs(more_total_charges - 1.2 * more_restaurant_bills)) < 1e-6

True

The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

#### Task 12 📍

What was the sum of all the bills in `more_restaurant_bills`, *including tips*?


In [50]:
sum_of_bills = sum(more_total_charges) # SOLUTION
sum_of_bills

1795730.0640000193

In [51]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct values."
failure_message: "❌ You don't seem to have the correct values. Try using the sum function."
""" # END TEST CONFIG

round(sum_of_bills, 2) == 1795730.06

True

#### Task 13 📍

The powers of 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  (For example, you may have noticed that storage on smartphones or USBs come in powers of 2, like 16 GB, 32 GB, or 64 GB.)  Use `np.arange` and the exponentiation operator `**` to compute the first 30 powers of 2, starting from $2^0$ up to and including $2^{29}$.

*Hint 1:* `np.arange(1, 2**30, 1)` creates an array with $2^{30}$ elements and **will crash your kernel**.

*Hint 2:* Part of your solution will involve `np.arange`, but your array shouldn't have more than 30 elements.


In [52]:
powers_of_2 = 2 ** np.arange(30) # SOLUTION
powers_of_2

array([        1,         2,         4,         8,        16,        32,
              64,       128,       256,       512,      1024,      2048,
            4096,      8192,     16384,     32768,     65536,    131072,
          262144,    524288,   1048576,   2097152,   4194304,   8388608,
        16777216,  33554432,  67108864, 134217728, 268435456, 536870912])

In [53]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct values."
failure_message: "❌ You don't seem to have the correct values. Try evaluating 2 with an exponent that is an array of values."
""" # END TEST CONFIG

all(powers_of_2 == 2 ** np.arange(30))

True

## Creating Tables

An array is useful for describing a single attribute of each element in a collection. For example, let's say our collection is all US States. Then an array could describe the land area of each state. 

Tables extend this idea by containing multiple arrays, each one describing a different attribute for every element of a collection. In this way, tables allow us to not only store data about many entities but to also contain several kinds of data about each entity.

For example, in the cell below we have two arrays. The first one, `population_amounts`, was defined above in section 2.2 and contains the world population in each year (estimated by the US Census Bureau). The second array, `years`, contains the years themselves. These elements are in order, so the year and the world population for that year have the same index in their corresponding arrays.

Run the following cell.

In [54]:
years = np.arange(1950, 2015+1)
print("Population column:", population_amounts)
print("Years column:", years)

Population column: [2557628654 2594939877 2636772306 2682053389 2730228104 2782098943
 2835299673 2891349717 2948137248 3000716593 3043001508 3083966929
 3140093217 3209827882 3281201306 3350425793 3420677923 3490333715
 3562313822 3637159050 3712697742 3790326948 3866568653 3942096442
 4016608813 4089083233 4160185010 4232084578 4304105753 4379013942
 4451362735 4534410125 4614566561 4695736743 4774569391 4856462699
 4940571232 5027200492 5114557167 5201440110 5288955934 5371585922
 5456136278 5538268316 5618682132 5699202985 5779440593 5857972543
 5935213248 6012074922 6088571383 6165219247 6242016348 6318590956
 6395699509 6473044732 6551263534 6629913759 6709049780 6788214394
 6866332358 6944055583 7022349283 7101027895 7178722893 7256490011]
Years column: [1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963 1964
 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977 1978 1979
 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993 1994
 

Suppose we want to answer this question:

> In which year did the world's population cross 6 billion?

You could technically answer this question just from staring at the arrays, but it's a bit convoluted, since you would have to count the position where the population first crossed 6 billion, then find the corresponding element in the years array. In cases like these, it might be easier to put the data into a *`Table`*, a 2-dimensional type of dataset. 

The expression below:

- creates an empty table using the expression `Table()`,
- adds two columns by calling `with_columns` with four arguments,
- assigns the result to the name `population`, and finally
- evaluates `population` so that we can see the table.

The strings `"Year"` and `"Population"` are column labels that we have chosen. The names `population_amounts` and `years` were assigned above to two arrays of the **same length**. The function `with_columns` (you can find the documentation [here](http://data8.org/datascience/tables.html)) takes in alternating strings (to represent column labels) and arrays (representing the data in those columns). The strings and arrays are separated by commas.

Run the following cell.

In [55]:
population = Table().with_columns(
    "Population", population_amounts,
    "Year", years
)
population

Population | Year
2557628654 | 1950
2594939877 | 1951
2636772306 | 1952
2682053389 | 1953
2730228104 | 1954
2782098943 | 1955
2835299673 | 1956
2891349717 | 1957
2948137248 | 1958
3000716593 | 1959
... (56 rows omitted)

Now the data is combined into a single table! It's much easier to parse this data. If you need to know what the population was in 1959, for example, you can tell from a single glance.

### Task 14 📍

In the cell below, we've created 2 arrays. Using the steps above, assign `top_10_movies` to a table that has two columns called "Rating" and "Name", which hold `top_10_movie_ratings` and `top_10_movie_names` respectively.


In [56]:
top_10_movie_ratings = make_array(9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8)
top_10_movie_names = make_array(
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)')

top_10_movies = Table().with_columns("Rating", top_10_movie_ratings, "Name", top_10_movie_names) # SOLUTION

# We've put this next line here 
# so your table will get printed out 
# when you run this cell.
top_10_movies

Rating | Name
9.2    | The Shawshank Redemption (1994)
9.2    | The Godfather (1972)
9      | The Godfather: Part II (1974)
8.9    | Pulp Fiction (1994)
8.9    | Schindler's List (1993)
8.9    | The Lord of the Rings: The Return of the King (2003)
8.9    | 12 Angry Men (1957)
8.9    | The Dark Knight (2008)
8.9    | Il buono, il brutto, il cattivo (1966)
8.8    | The Lord of the Rings: The Fellowship of the Ring (2001)

In [57]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you made a Table."
failure_message: "❌ You don't seem to have made a Table. Try using Table() as a starting point."
""" # END TEST CONFIG

type(top_10_movies) == tables.Table

True

In [58]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the right values in the right order."
failure_message: "❌ You don't seem to have the rigths values in the right order."
""" # END TEST CONFIG

top_10_movies.select('Rating', 'Name').sort('Name').column('Name').item(3) == "Schindler's List (1993)"

True

### Loading a Table From a File

In most cases, we aren't going to go through the trouble of typing in all the data manually. Instead, we load them in from an external source, like a data file. There are many formats for data files, but CSV ("comma-separated values") is the most common.

`Table.read_table(...)` takes one argument (a path to a data file in string format) and returns a table.  

#### Task 15 📍

`imdb.csv` contains a table of information about the 250 highest-rated movies on IMDb.  Load it as a table called `imdb`.

(You may remember working with this table in Lab 2!)


In [59]:
imdb = Table.read_table('imdb.csv') # SOLUTION
imdb

Votes  | Rating | Title                 | Year | Decade
88355  | 8.4    | M                     | 1931 | 1930
132823 | 8.3    | Singin' in the Rain   | 1952 | 1950
74178  | 8.3    | All About Eve         | 1950 | 1950
635139 | 8.6    | Léon                  | 1994 | 1990
145514 | 8.2    | The Elephant Man      | 1980 | 1980
425461 | 8.3    | Full Metal Jacket     | 1987 | 1980
441174 | 8.1    | Gone Girl             | 2014 | 2010
850601 | 8.3    | Batman Begins         | 2005 | 2000
37664  | 8.2    | Judgment at Nuremberg | 1961 | 1960
46987  | 8      | Relatos salvajes      | 2014 | 2010
... (240 rows omitted)

In [60]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you made a Table."
failure_message: "❌ You don't seem to have made a Table. Try using Table.read_table() with the csv file in a string format as the arugment."
""" # END TEST CONFIG

type(imdb) == tables.Table

True

In [61]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ You have the correct number of rows."
failure_message: "❌ You don't seem to have the correct number of rows."
""" # END TEST CONFIG

imdb.num_rows == 250

True

In [62]:
imdb.select('Votes', 'Rating', 'Title', 'Year', 'Decade').sort(0).take(range(2,5))

Votes | Rating | Title                       | Year | Decade
31003 | 8.1    | Le salaire de la peur       | 1953 | 1950
32385 | 8      | La battaglia di Algeri      | 1966 | 1960
35983 | 8.1    | The Best Years of Our Lives | 1946 | 1940

Where did `imdb.csv` come from? Take a look at [this lab's folder](./). You should see a file called `imdb.csv`.

Open up the `imdb.csv` file in that folder and look at the format. What do you notice? The `.csv` filename ending says that this file is in the [CSV (comma-separated value) format](http://edoceo.com/utilitas/csv-file-format).

## More Table Operations!

Now that you've worked with arrays, let's add a few more methods to the list of table operations that you saw in Lab 2.

### `column`

`column` takes the column name of a table (in string format) as its argument and returns the values in that column as an **array**. 

In [63]:
# Returns an array of movie names
top_10_movies.column('Name')

array(['The Shawshank Redemption (1994)', 'The Godfather (1972)',
       'The Godfather: Part II (1974)', 'Pulp Fiction (1994)',
       "Schindler's List (1993)",
       'The Lord of the Rings: The Return of the King (2003)',
       '12 Angry Men (1957)', 'The Dark Knight (2008)',
       'Il buono, il brutto, il cattivo (1966)',
       'The Lord of the Rings: The Fellowship of the Ring (2001)'],
      dtype='<U56')

### `take`

The table method `take` takes as its argument an array of numbers.  Each number should be the index of a row in the table.  It returns a **new table** with only those rows. 

You'll usually want to use `take` in conjunction with `np.arange` to take the first few rows of a table.

In [64]:
# Take first 5 movies of top_10_movies
top_10_movies.take(np.arange(0, 5, 1))

Rating | Name
9.2    | The Shawshank Redemption (1994)
9.2    | The Godfather (1972)
9      | The Godfather: Part II (1974)
8.9    | Pulp Fiction (1994)
8.9    | Schindler's List (1993)

The next three questions will give you practice with combining the operations you've learned in this lab and the previous one to answer questions about the `population` and `imdb` tables. First, check out the `population` table from section 2.

In [65]:
# Run this cell to display the population table.
population

Population | Year
2557628654 | 1950
2594939877 | 1951
2636772306 | 1952
2682053389 | 1953
2730228104 | 1954
2782098943 | 1955
2835299673 | 1956
2891349717 | 1957
2948137248 | 1958
3000716593 | 1959
... (56 rows omitted)

### Task 16 📍

Check out the `population` table from section 2 of this lab.  Compute the year when the world population first went above 6 billion. Assign the year to `year_population_crossed_6_billion`.


In [66]:
year_population_crossed_6_billion = population.where('Population', are.above_or_equal_to(6*10**9)).column('Year').item(0) # SOLUTION
year_population_crossed_6_billion

1999

In [67]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the right data type for the variable."
failure_message: "❌ You don't seem like you have the right data type for the variable. Try using .column and .item"
""" # END TEST CONFIG

type(year_population_crossed_6_billion) == int or type(year_population_crossed_6_billion) == np.int32

True

In [68]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct value."
failure_message: "❌ You don't seem like you have the correct value. Make sure you used the where method and check your predicate."
""" # END TEST CONFIG

year_population_crossed_6_billion == 1999

True

### Task 17 📍

Find the average rating for movies released before the year 2000 and the average rating for movies released in the year 2000 or after for the movies in `imdb`.

*Hint*: Think of the steps you need to do (take the average, find the ratings, find movies released in 20th/21st centuries), and try to put them in an order that makes sense.


In [69]:
before_2000 = np.mean(imdb.where('Year', are.below(2000)).column('Rating')) # SOLUTION
after_or_in_2000 = np.mean(imdb.where('Year', are.above_or_equal_to(2000)).column('Rating')) # SOLUTION
print("Average before 2000 rating:", before_2000)
print("Average after or in 2000 rating:", after_or_in_2000)

Average before 2000 rating: 8.2783625731
Average after or in 2000 rating: 8.23797468354


In [70]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct value for before_2000."
failure_message: "❌ It doesn't seems like you have the correct value for before_2000. Try using np.mean, where, are.below, and column in your solution"
""" # END TEST CONFIG

abs(before_2000 - 8.2783625730994146) < 1e-5

True

In [71]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ It seems like you have the correct value for after_or_in_2000."
failure_message: "❌ It doesn't seems like you have the correct value for after_or_in_2000. Try using np.mean, where, are.above_or_equal_to, and column in your solution"
""" # END TEST CONFIG

abs(after_or_in_2000 - 8.2379746835443033) < 1e-5

True

### Task 18 📍

Here's a challenge: Find the number of movies that came out in *even* years.

*Hint:* The operator `%` computes the remainder when dividing by a number.  So `5 % 2` is 1 and `6 % 2` is 0.  A number is even if the remainder is 0 when you divide by 2.

*Hint 2:* `%` can be used on arrays, operating elementwise like `+` or `*`.  So `make_array(5, 6, 7) % 2` is `array([1, 0, 1])`.

*Hint 3:* Create a column called "Year Remainder" that's the remainder when each movie's release year is divided by 2.  Make a copy of `imdb` that includes that column (`imdb.with_column(...)` returns a new table).  Then use `where` to find rows where that new column is equal to 0.  Then use `num_rows` to count the number of such rows.

*Note:* These steps can be chained in one single statement, or broken up across several lines with intermediate names assigned. You’re always welcome to break down problems however you wish!


In [72]:
num_even_year_movies = imdb.with_column("Year Remainder", imdb.column("Year") % 2).where("Year Remainder", are.equal_to(0)).num_rows # SOLUTION
num_even_year_movies

127

In [73]:
""" # BEGIN TEST CONFIG
hidden: false
points: 1
success_message: "✅ You have the correct value."
failure_message: "❌ It doesn't seem that you have the correct value. Trying using % 2 to help you identify even numbers. The remainder should be 0."
""" # END TEST CONFIG

num_even_year_movies == 127

True

# Submit your Lab to Canvas

Once you have finished working on the lab questions, prepare to submit your work in Canvas by completing the following steps.

1. Double-check that you have run the code cell near the end of the notebook that contains the command "grader.check_all()". This command will run all of the run tests on all your responses to the code questions.
2. Double-check that you have responded to all the open-response questions.
3. Select the menu items "File", "Save and Export Notebook As...", and "HTML (.html)" in the notebook's Toolbar to download an HTML version of this notebook file.
4. In the related Canvas Assignment, click Start Assignment or New Attempt to upload the downloaded HTML file.